### PR TITLE
Restructure stream and budget manager tests

### PR DESCRIPTION
**Pull request by AlexanderBenz**
See #303

**Additional context**
Includes test for stream and budget manager

### CODE REVIEW

**Review by mherde — CHANGES_REQUESTED** [review]

File: skactiveml/tests/template_query_strategy.py [hunk 1]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
                     pass
 
 
+class TemplateSingleAnnotatorStreamQueryStrategy(TemplateQueryStrategy):
+    def setUp(
+        self,
+        qs_class,
+        init_default_params,
+        query_default_params_clf=None,
+    ):
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:640]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 2]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+            init_default_params,
+            query_default_params_clf,
+            None,
+        )
+        self.update_params = {
+            "candidates": [[]],
+            "queried_indices": [],
+        }
+
+    def test_query_param_clf(self, test_cases=None):
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:652]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 3]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+        query_params = inspect.signature(self.qs_class.query).parameters
+        if "X" in query_params:
+            test_cases = [] if test_cases is None else test_cases
+            test_cases += [
+                ("string", (ValueError, TypeError)),
+                (Dummy, (ValueError, TypeError)),
+            ]
+            self._test_param("query", "X", test_cases)
+
+            for exclude_clf, exclude_reg, query_params in [
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:687]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 4]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+                y = self.query_default_params_clf["y"]
+                test_cases = [(y, None), (np.vstack([y, y]), ValueError)]
+                self._test_param("query", "y", test_cases, exclude_reg=True)
+
+                for ml, classes, t, err in [
+                    (np.nan, [1.0, 2.0], float, None),
+                    (0, [1, 2], int, None),
+                    (None, [1, 2], object, None),
+                    (None, ["A", "B"], object, None),
+                    ("", ["A", "B"], str, None),
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:720]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 5]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+                        replace_init_params=replace_init_params,
+                        replace_query_params=replace_query_params,
+                        exclude_reg=True,
+                    )
+
+    def test_query_param_candidates(self, test_cases=None):  # TODO more cases
+        test_cases = [] if test_cases is None else test_cases
+
+        for exclude_clf, exclude_reg, query_params in [
+            (False, True, self.query_default_params_clf),
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:750]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 6]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+        ]:
+            if query_params is not None:
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:752]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 7]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+            test_cases += [
+                (np.nan, (ValueError, TypeError)),
+                (Dummy, (ValueError, TypeError)),
+                (None, None),
+            ]
+            self._test_param("query", "sample_weight", test_cases)
+
+            for exclude_clf, exclude_reg, query_params in [
+                (False, True, self.query_default_params_clf),
+                # (True, False, self.query_default_params_reg),
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:781]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 8]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+                        "sample_weight",
+                        test_cases,
+                        replace_query_params=replace_query_params,
+                        exclude_clf=exclude_clf,
+                        exclude_reg=exclude_reg,
+                    )
+
+    def test_query(
+        self,
+        expected_output=None,
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:801]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 9]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+        expected_utilities=None,
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:802]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 10]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+            raise ValueError(
+                "Test need to override expected_output and expected_utilities"
+            )
+        init_params = deepcopy(self.init_default_params)
+        init_params["random_state"] = np.random.RandomState(0)
+        qs = self.qs_class(**init_params)
+        qs2 = self.qs_class(**init_params)
+        X = np.array([[0, 0], [0, 1], [1, 0], [1, 1], [0.75, 0.75]])
+        y = np.array([0, 0, 1, 1, 1])
+        candidat = np.array([[0.5, 0.5]])
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:815]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 11]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+        X = [[0, 0], [0, 1], [1, 0], [1, 1], [0.75, 0.75], [0.5, 0.5]]
+        y_true = [0, 0, 1, 1, 1, 0]
+        query_default_params1 = deepcopy(self.query_default_params_clf)
+        query_params = inspect.signature(self.qs_class.query).parameters
+        utilities = []
+        X_queue = []
+        y_queue = []
+        qs_outputs = []
+        for i, x in enumerate(X):
+            if "clf" in query_params and i > 0:
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py:916]: Done

File: skactiveml/tests/template_query_strategy.py [hunk 12]
```diff
@@ -630,6 +631,351 @@ def test_query_candidate_variation(self):
...
+                id1, u1 = qs.query(**query_params)
+                id2, u2 = qs.query(**query_params)
+
+                self.assertEqual(len(u1), len(query_params["candidates"]))
+                np.testing.assert_array_equal(id1, id2)
+                np.testing.assert_allclose(u1, u2)
+
+    def test_update_param_candidates(self, test_cases=None):
+        test_cases = [] if test_cases is None else test_cases
+        test_cases += [(Dummy, TypeError), ([[]], None)]
```
[the remark this answers is not in the document]
Comment (AlexanderBenz) [reply skactiveml/tests/template_query_strategy.py]: Done

**Review by mherde — CHANGES_REQUESTED** [review]

File: skactiveml/tests/template_query_strategy.py [hunk 13]
```diff
@@ -630,6 +634,440 @@ def test_query_candidate_variation(self):
...
+                        replace_query_params=replace_query_params,
+                        exclude_clf=exclude_clf,
+                        exclude_reg=exclude_reg,
+                    )
+
+    def test_query(
+        self,
+        expected_output,
+        expected_utilities,
+        budget_manager_param_dict=None,
```
Comment (mherde) [inline skactiveml/tests/template_query_strategy.py:843]:
```suggestion
        budget_manager_param_dict=None,
        X=None,
        y=None,
        candidate=None,
        queried_indices=None
```
Add these parameters as optional ones. If they are set to `None`, we use your current setup as default. As a default case, generating the candidate with a fixed random state would also be better to have better variability.

File: skactiveml/tests/template_query_strategy.py [hunk 14]
```diff
@@ -630,6 +634,440 @@ def test_query_candidate_variation(self):
...
+                self.assertEqual(len(qs_output2), len(qs_output))
+            else:
+                np.testing.assert_array_equal(
+                    np.array(expected_output), np.array(qs_output)
+                )
+                np.testing.assert_array_equal(
+                    np.array(qs_output2), np.array(qs_output)
+                )
+                # self.assertEqual(expected_output, qs_output)
+                # self.assertEqual(qs_output2, qs_output)
```
Comment (mherde) [inline skactiveml/tests/template_query_strategy.py:926]:
```suggestion

```

File: skactiveml/tests/template_query_strategy.py [hunk 15]
```diff
@@ -630,6 +634,440 @@ def test_query_candidate_variation(self):
...
+                self.assertEqual(len(expected_output), len(qs_output))
+                self.assertEqual(len(qs_output2), len(qs_output))
+            else:
+                np.testing.assert_array_equal(
+                    np.array(expected_output), np.array(qs_output)
+                )
+                np.testing.assert_array_equal(
+                    np.array(qs_output2), np.array(qs_output)
+                )
+                # self.assertEqual(expected_output, qs_output)
```
Comment (mherde) [inline skactiveml/tests/template_query_strategy.py:925]:
```suggestion
```

File: skactiveml/tests/template_query_strategy.py (hunk 13 above)
Comment (mherde) [reply skactiveml/tests/template_query_strategy.py:843]: Add comments to have a better overview of this long method.

File: skactiveml/tests/template_query_strategy.py [hunk 16]
```diff
@@ -630,6 +634,440 @@ def test_query_candidate_variation(self):
...
+            )
+            call_func(
+                qs2.update,
+                candidates=X,
+                queried_indices=[0, 1, 2, 3, 4],
+                budget_manager_param_dict=budget_manager_param_dict,
+            )
+            qs_output, utilities = qs.query(**query_default_params)
+            qs_output2 = []
+            utilities2 = []
```
Comment (mherde) [inline skactiveml/tests/template_query_strategy.py:910]:
```suggestion
```

File: skactiveml/tests/template_query_strategy.py [hunk 17]
```diff
@@ -630,6 +634,440 @@ def test_query_candidate_variation(self):
...
+                budget_manager_param_dict=budget_manager_param_dict,
+            )
+            call_func(
+                qs2.update,
+                candidates=X,
+                queried_indices=[0, 1, 2, 3, 4],
+                budget_manager_param_dict=budget_manager_param_dict,
+            )
+            qs_output, utilities = qs.query(**query_default_params)
+            qs_output2 = []
```
Comment (mherde) [inline skactiveml/tests/template_query_strategy.py:909]:
```suggestion
```

**Review by mherde — CHANGES_REQUESTED** [review]

File: skactiveml/tests/template_budget_manager.py [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+        init_params = deepcopy(self.init_default_params)
+        init_params_list = inspect.signature(self.bm_class.__init__).parameters
+        if "random_state" in init_params_list:
+            init_params["random_state"] = np.random.RandomState(0)
+        bm = self.bm_class(**init_params)
+        bm2 = self.bm_class(**init_params)
+        bm2_outputs = []
+        utilities_update = []
+        utilities = np.array([0.2, 0.6, 0.8, 0.9, 0.1])
+        candidat = np.array([0.3])
```
Comment (mherde) [inline skactiveml/tests/template_budget_manager.py:246]:
```suggestion
        candidate = np.array([0.3])
```

File: skactiveml/tests/template_budget_manager.py [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+                    self.assertRaises(
+                        err, bm.query_by_utility, **query_by_utility_params
+                    )
+                else:
+                    func = getattr(bm, test_func)
+                    self.assertRaises(err, func, **update_params)
+
+    def test_query_by_utility(
+        self,
+        expected_output=None,
```
Comment (mherde) [inline skactiveml/tests/template_budget_manager.py:187]:
```suggestion
        expected_output,
```

**Review by mherde — APPROVED** [review]

**Pull request merged**